### DISCUSSION TITLE
Physik Debug - Draw Raycast before simulate, instead of after

### DISCUSSION THREAD

**Discussion by TheMarkoD** [JirkaDellOro/FUDGE#257] [General]
@JirkaDellOro 
Habe es mir mal angeschaut.
Es stimmt, intuitiver wäre es wenn die Raycasts gezeichnet werden würden, wenn man sie vor dem `physics.simulate()` macht und simulate die Physik quasi abschließt für den Frame.

Ich stoße dabei auf folgendes Problem. Das Debug besteht aus eigenen Buffern, die dann in den Kontext von Fudge gezeichnet werden. Die Buffer werden in der `simulate()` geleert mittels `begin()`, der debugDraw von Oimo befüllt aus der dann simulierten Welt den Buffer. Das tatsächliche Zeichnen ist dann im Render z. 153 das `end()`, da werden die gefüllten Buffer gezeichnet.

Wenn wir den Raycast quasi nach der Simulate ausführen ist er 1 Frame verschoben 
(wobei das kein Problem darstellt, 3-5 Frames gelten als nicht wahrnehmbar, zumal der Strahl nur 1 Frame existiert). Da aber die Intuition ist zuerst zu Casten und dann zu simulaten, wäre eine Umstellung gut.

Aber dafür muss der Zeichenbefehl für den Raycast nach dem **leeren** kommen. Das lässt der Programmcode so nicht zu, ich schlage daher eine Zwischenspeicherung aller Zeichenbefehle vor. Und im Schritt, simulate(), wird dann eine drawDebugRays() aufgerufen, die für jeden Strahl im Array den Zeichenbefehl ausführt, nach der Leerung. Anschließend wird das Array geleert.

Dadurch ist sichergestellt, dass Strahlen gezeichnet werden die sowohl vor als nach der Simulate ausgeführt werden.
 
Idee: 
Denkbar wäre da die Erweiterung in diesem Array eine Dauer bzw. ein ExpirationDate mit einzuspeichern, d.h. es wird nicht das gesamte debugRays[] gelöscht, sondern nur die abgelaufenen. Das gäbe die Möglichkeit, wie in Unity einen Strahl für eine Dauer sichtbar zu lassen, auch wenn die Physik nicht mehr damit arbeitet, für verbessertes Debugging eines z.B. Schusses, der ja nur in einem Frame und nicht kontinuierlich ausgeführt wird.

**TL;DR;**
- Aufgrund der Reihenfolge des Buffers, muss bisher der Raycast nach dem simulate() erfolgen. Der Leervorgang des DebugDraw Buffers muss aber so bestehen bleiben.

- Mittels eines Zwischenspeicher Arrays (mehr Overhead), könnte aber Möglichkeit geschaffen werden, die Raycasting vor/nach dem simulate() frei erlaubt. Zudem wäre es möglich den Raycast für eine gegebene Dauer, immer zu zeichnen (z.B. Debugging eines Einzelschusses, analog zu Unity)

- Habe das jetzt nur mal getestet, wenn das aber gewünscht ist, kann ich das mal ordentlich reincoden und irgendwie pushen.

**Comment by JirkaDellOro** [comment]
Mein erster Vorschlag wäre ähnlich: dass man Extra-Buffer für die Rays macht, diese am Ende des simulate zeichnet und danach löscht. Dann sind sie im nächsten Frame leer. 

Vielleicht könntest Du das aber auch einfach grundsätzlich bei simulate machen, dass Du die Zeichenbuffer am Ende löschst und nicht "erst" am Anfang? Dann wären sie also am Ende von simulate leer und man könnte sie im nächsten Frame noch vor simulate schon z.B. mit Rays befüllen?

Das mit der Dauer erscheint mir jetzt zu viel Aufwand für den Effekt.

**Comment by TheMarkoD** [comment]
Könnte man andenken, ich werde mal etwas coden, wenns reinpasst. 
Dürfte ja keine allzu hohe Prio gerade haben, die Raycasts nach dem Simulate zu setzen.

Ich finde den Aufwand für den Effekt eigentlich angemessen, habe ich innerhalb der Thesis auch noch nicht gedacht. 

Aber wenn ich mich zurück an meine Bachelor Thesis erinnere, wie oft ich in einem Shooter einen Schuss debuggen muss, der üblicherweise per Klick in einem Frame gestartet wird, aber dann auch beobachtet werden sollte, dann halte ich es für angemessen. Alle anderen Buffer werden gelöscht, es wäre nur der "neu" angedachte Raycast Buffer, der eben zeitlich relevant gelöscht wird. 
Ist für Nutzer denke ich angenehmer für Raycasts in den Settings eine allgemeine Löschzeit anzugeben, anstatt für jeden Ray, wie in Unity. Aber insgesamt eine angenehmere Lösung, als eine schon programmierte Gameplay Mechanik, wie einen Schuss, aus einer fixen Eingabe, in eine kontinuierliche Eingabe zu verlagern. 

**Edit** 
Der Buffer kann, nicht im Simulate geleert werden. Sondern sinnvollerweise nur zu Beginn. 
Simulate löscht den Buffer, befüllt ihn (bis auf Raycasts) aus dem DebugDraw und gezeichnet wird erst im Rendering. Wenn Simulate am Ende ihrer Ausführung den Buffer löschen würde, dann wäre er ja im Rendering nicht mehr da. Die Reihenfolge passt. Es muss eben nur noch einen Buffer für Raycasts geben, da schaue ich eben, dass ich das ggf. einbaue.

**Reply by JirkaDellOro** [reply]
> Wenn Simulate am Ende ihrer Ausführung den Buffer löschen würde, dann wäre er ja im Rendering nicht mehr da.   
klar, zu kurz gedacht.

Für den Effekt ist es auch denkbar, dass der Creator selbst den Ray länger anzeigen lässt und nur in einem Frame die Auswertung vornimmt. Ich finde es jetzt schon super, dass man überhaupt tatsächlich den Ray sieht und nicht einen anderen, den man wie bei Unity parallel bauen muss.

**Comment by JirkaDellOro** [comment]
Ich habe das mal von Issues in Discussions geräumt, auch weil das jetzt endlich möglich ist bei Github. Schönes Feature. Aber auch weil mir zum Thema noch eine Idee kam:

Wie wäre es, wenn wir simulate und draw separieren? Und stattdessen der Creator selbst angibt, wenn er die Physik sehen will.
```typescript
    viewport.draw();
    viewport.drawPhysics();
```
Wobei er noch Parameter angeben kann, z.B. was genau dargestellt werden soll. Der Viewport übernimmt dann die Kameradefinition für die Physik etc. und Render.draw würde dann wohl zusammenschrumpeln von 
```typescript
    public static draw(_cmpCamera: ComponentCamera): void {
      // TODO: Move physics rendering to RenderPhysics extension of RenderManager
      if (Physics.world && Physics.world.mainCam != _cmpCamera)
        Physics.world.mainCam = _cmpCamera; //DebugDraw needs to know the main camera beforehand, _cmpCamera is the viewport camera. | Marko Fehrenbach, HFU 2020

      // TODO: check physics
      if (!Physics.settings || Physics.settings.debugMode != PHYSICS_DEBUGMODE.PHYSIC_OBJECTS_ONLY) { //Give users the possibility to only show physics displayed | Marko Fehrenbach, HFU 2020
        Render.drawList(_cmpCamera, this.nodesSimple);
        Render.drawListAlpha(_cmpCamera);
      }
      if (Physics.settings && Physics.settings.debugDraw == true) {
        Physics.world.debugDraw.end();
      }
    }
```
zu
```typescript 
    public static draw(_cmpCamera: ComponentCamera): void {
        Render.drawList(_cmpCamera, this.nodesSimple);
        Render.drawListAlpha(_cmpCamera);
    }
```

Spricht etwas dagegen? Ich schau mal, ob ich das ausprobieren kann.

**Reply by JirkaDellOro** [reply]
ganz so weit auftrennen geht nicht, da der Viewport das Gesamtbild am Ende auf den Canvas klebt. Render.draw sieht aber jetzt aus wie oben angekündigt und hat keine Bezüge auf Physics mehr, dagegen findet sich in Viewport nun das:

```typescript
      Render.clear(this.camera.clrBackground);
      if (Physics.settings?.debugMode != PHYSICS_DEBUGMODE.PHYSIC_OBJECTS_ONLY)
        Render.draw(this.camera);
      if (Physics.settings?.debugDraw) {
        Physics.world.draw(this.camera);
      }
```

Und in Physics sieht es so aus  

```typescript
    public simulate(_deltaTime: number = 1 / 60): void {
      if (this.jointList.length > 0)
        this.connectJoints(); //Connect joints if anything has happened between the last call to any of the two paired rigidbodies
      if (Time.game.getScale() != 0) //If time is stopped do not simulate to avoid misbehaviour
        Physics.world.oimoWorld.step(_deltaTime * Time.game.getScale());  //Update the simulation by the given deltaTime and the Fudge internal TimeScale
    }

    public draw(_cmpCamera: ComponentCamera): void {
      Physics.world.mainCam = _cmpCamera;
      Physics.world.debugDraw.begin();  //Updates info about the current projection, resetting the points/lines/triangles that need to be drawn from debug
      Physics.world.oimoWorld.debugDraw(); //Filling the physics world debug informations into the debug rendering handler
      Physics.world.debugDraw.end();
    }
```

So scheint mir alles aufgeräumter und besser verortet.

**Reply by JirkaDellOro** [reply]
Ist in devJirka

**Reply by JirkaDellOro** [reply]
Ich habe debugDraw.begin und end in clearBuffers und drawBuffers umbenannt, dann wird deutlich, was da passiert. Physics.draw habe ich umsortiert, sodass wie oben angeregt, am Ende des draw ein clear kommt

```typescript
    public draw(_cmpCamera: ComponentCamera): void {
      Physics.world.debugDraw.getDebugModeFromSettings();
      Physics.world.mainCam = _cmpCamera;
      Physics.world.oimoWorld.debugDraw(); //Filling the physics world debug informations into the debug rendering handler
      Physics.world.debugDraw.drawBuffers();
      Physics.world.debugDraw.clearBuffers();  //Updates info about the current projection, resetting the points/lines/triangles that need to be drawn from debug
    }
```

Damit kann nun der Ray irgendwann gezeichnet werden und die Darstellung der Meshes und der Collider laufen nicht mehr auseinander. Das erste clearBuffer kommt in initializeBuffers von PhysicsDebugDraw, das scheint gut dahin zu passen und die Buffer sind dann auf jeden Fall angelegt und startklar.

Einwände?

**Reply by TheMarkoD** [reply]
Nein keine Einwände, sieht soweit sehr gut aus und es scheint auch bei alle Tests noch alles zu funktionieren.

Wenn ich das richtig sehe, haben diese Änderungen auch zu Folge, dass die Position des Raycasts egal ist, also ob vor/nach `simulate()`. 
Jetzt kann natürlich überlegen ob es nötig ist länger als ein Frame den Ray zu debuggen. Fürs erste kann denke ich gut damit gearbeitet werden, wenn er kontinuierlich ausgeführt werden muss um gezeichnet zu werden, wenn das so kommuniziert ist.
